wordpress-mobile/WordPress-Android
# [UITests] Fix crash with Arabic locale

Fixes parts of #12601

## Context

When running the UI Tests used to generate the screenshots (`WPScreenshotTest`), the tests crashes when using the `ar` locale (Arabic).

This is due to the mocks returned by WireMock, which generate an unparsable date, throwing an exception when decoding the JSON from those stubs 

### Root cause of the issue

We use WireMock's "templating" features to define our mocks, allowing us to inject dynamic dates. For that, we use [the {{now}} helper](http://wiremock.org/docs/response-templating/#date-and-time-helpers) in our mocks like [in this template mock here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/libs/mocks/WordPressMocks/src/main/assets/mocks/mappings/wpcom/stats/stats_top-posts.json#L9)

The issue with this is that WireMock uses the **current user locale** to render the date when transforming this `{{now}}` tag. This means that in any locale using Roman alphabet/dates/number system this may work, but for locales like arabic, the date once rendered by WireMock's `{{now}}` tag looks like some garbage text (character encoding also getting in the way of interpreting those valid arabic characters strangely), which then generates an `Unparsable date` exception while trying to decode the mocked JSON.

### Fixing the locale issue in mocks

This was a hard one to track. Ultimately, the solution boiled down to:

 * Registering a new `{{fnow}}` tag in `WireMockRule` that will do something similar to the `{{now}}` official tag, but use the `en_US_POSIX` locale instead
   * That work had already been started by @jkmassel a while ago; but the tag never got used in the stubs after all, and the parts of the implementation from back then that were still present in the codebase actually crashed WireMock with a `ClassCastException`.
   * I thus had to update the method signature in `UnlocalizedDateHelper` (implementation for our custom `{{fnow}}`  tag) to take an `Object` for the context instead of a `Date`. That's because `{{fnow}}` tag doesn't take a `Date` input but instead generates an output value independent of any input, so there's no reason we should expect the input context to be a `Date`, and in fact we won't even use the template's context as we won't need any input.
   * After that fix I was able to finally make a `{{fnow}}` tag in a JSON template trigger that `apply(…)` method… But then the method itself crashed; this was because we were trying to build a `Locale` using the code `"US"`, while that's not a valid code for a locale
    * Besides, as per ISO standards, the right locale to use when parsing "internet dates" (aka locale-agnostic dates) is **`en_US_POSIX`**, not `en_US` anyway.

Once I had the `{{fnow}}` registered and its implementation fixed, I was able to change all the instances of `{{now…` in the mock templates into `{{fnow…` instead, finally fixing the core of the stubs issue with non-roman (and arabic in particular) locales

_(Once again, the longest debugging sessions often lead to the smallest fixes… 😅)_

## Bonus: Fixing a small point forgotten in past locale-related PR

When I implemented the PR to fix the Indonesian locale code used for screenshots, I forgot to rename the `fastlane/playstoreres/id` folder to `in`. I realised that when testing other locales to validate the current PR, so I thought it would be worth including there too.

This change is not strictly mandatory as the proper resources will be downloaded in `fastlane/playstoreres/in` next time you run `download_promo_strings` anyway, but I thought it was cleaner to have the right folder name committed into git to avoid keeping any legacy naming around.

## To test:

* Launch an emulator in AndroidStudio, switch the locale to Arabic, then run the `WPScreenshotTest` in AndroidStudio
* Alternatively, edit the `WPScreenshotTest.java` file on line 42 to change `LOCALE_TEST_RULE = new WPLocaleTestRule()` to `LOCALE_TEST_RULE = new WPLocaleTestRule("ar", "en")`, then run the test
* Verify that the test doesn't crash anymore at the end and that the parsing of the dates from the stub don't fail with an exception
